client/state/guided-tours/selectors/index.js

```diff
@@ -70,12 +71,11 @@ const getTourFromQuery = createSelector(
 	( state ) => {
 		const initial = getInitialQueryArguments( state );
 		const current = getCurrentQueryArguments( state );
-		const tourProps = [ 'tour', '_timestamp' ];
-		const { tour, _timestamp } =
-			current && current.tour ? pick( current, tourProps ) : pick( initial, tourProps );
+		const timestamp = getCurrentRouteTimestamp( state );
+		const tour = current && current.tour ? current.tour : initial.tour;
```

Comment:
I think that we can maybe abbreviate the ternary operator using optional chaining here.

Reply:
And perhaps some nullish coalescing too:

```suggestion
		const tour = current?.tour ?? initial.tour;
```

Reply:
I decided to do a `current.tour ?? initial.tour` coalescing, without checking the nullishness of `current` or `initial`. They are guaranteed to be at least `false`, which is a `Boolean` instance with field access possible, and we're going to migrate that to empty `{}` objects.

At other places, the results of `get*QueryArguments` are used without guards, too.

Reply:
Sounds good to me. Looks like we might need to update some tests, too.

Reply:
Just pushed a fix. Mock state was missing `state.route.query.current` and that's not what happens IRL 🙂

Reply:
Speaking of the default `{}` (and also for empty instances), I wonder why not use of `null` for those cases. To me `null` is synonymous to "intentional undefined" or "knowingly missing/empty value", and that's something we can apply uniformly across the app (as we also do here for `timestamp`, for example). An empty object apparently helps with further uses in selectors (not having to check every time - like the argument with "false" above), but semantically (or design-wise) speaking, I wish "empty" or "missing" or "none found" could be described uniformly at the state tree level, and `null` (or `undefined` as an initial value) fit that case well.

Reply:
For URL query strings, the `null` value would make sense only when the state value is truly unknown, when no navigation has happened yet. And that's an edge case that happens only for a very short period of time during Calypso boot, after creating the Redux store and before the initial route is handled after `page.start()` and the `ROUTE_SET` action is dispatched.

Similarly, the initial state of `state.route.path.current` should be `null` instead of the `''` empty string.

When the current URL doesn't have a query string, the correct value for `state.route.query.current` is an empty set (`{}`), not `null`. The state is known. Similarly, `window.location.search` is an empty string (`''`) in this case, and `new URLSearchParams( window.location.search )` will produce an object with zero entries.

So, in theory, it's true that `state.route.query.current` is initially `null` and that we should guard all field accesses, but it's not practical: in practice we're always on some URL.

Reply:
> For URL query strings, the null value would make sense only when the state value is truly unknown, when no navigation has happened yet. And that's an edge case that happens only for a very short period of time during Calypso boot, after creating the Redux store and before the initial route is handled after page.start() and the ROUTE_SET action is dispatched.

I would think `undefined` for that case, as you are describing something that has even been attempted to have a value yet. When we do navigate to a page and find no params, then we are aware of its emptiness - it is a fact, and we account for it. I don't think of `null` and `undefined` as the same - their intention is different (despite their presence being consequential or not).

Reply:
> I would think `undefined` for that case

The trouble here is that Redux doesn't like `undefined` state very much. It has a special meaning there: uninitialized state. And reducers use the `( state = initial, action ) => { ... }` idiom to initialize state. If a reducer returns `undefined`, Redux will log a warning.

That's why the "state is unknown, e.g., because hasn't been fetched from server yet" situation is usually represented as `null` in Calypso.

> When we do navigate to a page and find no params, then we are aware of its emptiness

I don't understand your point here: I'm recommending to represent this "emptiness" as an empty set, `{}`. Do we have an agreement on this?

Reply:
Gotcha on the Redux part. If we have a technical restriction/limitation, then there isn't much we can do.

> I don't understand your point here: I'm recommending to represent this "emptiness" as an empty set, {}. Do we have an agreement on this?

I am trying to entertain the alternative (and apologies if I am dragging this more than necessary). That we use `null` for this.

My point is/remains about uniformity i.e. it makes little difference if something is a string or a number or an object or a function, etc. I think the discrepancy between `undefined` and `null` do give us a uniform layer to describe "emptiness" that way. `undefined` for some form of initial state that we have not yet observed (akin to undeclared), and `null` for things that we are aware and explicitly empty out. 

> The state is known. Similarly, window.location.search is an empty string ('') in this case, and new URLSearchParams( window.location.search ) will produce an object with zero entries.

Speaking about uniformity, these two are probably not good examples. The `window.location.search` API treats "emptiness" as an empty string (`''`). `new URLSearchParams` wraps it in a `URLSearchParams` object, with an iterator that's empty. So emptiness here is not straightforward - although, it can be seen as a form of `undefined` or `null` once you interact with that API. 

```
const foo = new URLSearchParams( '' );

foo.values().next();
> { done: true, value: undefined }

foo.get( 'bar' );
> null
 ```

Of course, just for the sake of discussion. Apologies for dragging. I see your point about an empty object `{}` being valid.

Reply:
> Speaking about uniformity, these two are probably not good examples.

For me, the main benefit is that the `new URLSearchParams( window.location.search )` and `params.get( 'foo' )` operations will always work, and will work correctly. There are never runtime exceptions about the `URLSearchParams` constructor getting a wrong parameter or missing `.get` method on a non-object. Whether the URL has query params or not, the code is always the same and I don't need any guards or conditionals.

The empty `query = {}` object delivers the same benefit, and it's not semantically wrong.

Reply:
When I think of APIs I see them as an access layer over some data. The way that the data are structured internally is obscured to the consumer; she only sees the interface. We can build that interface to be consistent and deterministic like`URLSearchParams` (although it doesn't look as consistent when I get `undefined` when accessing the iterator value directly, while `null` via the getter - but irrespective). 

In our case, the consumer (component) interacts via a selector. For all we know the selector may implement `URLSearchParams` underneath. The data though is held in whatever structure we want. In this case, `null` when no data, an array of `tuples` or an `object` when data are present. It makes no difference. `getQueryArgs` will be consistent, it will work correctly. So will `getQueryArgsAsArray`, or `getQueryArgsAsObject`, etc. Selectors are the API we provide to components, not the state tree.

When the consumer wants to peak at the actual data, they will see nulls, and undefined, and whatever structure we implement. And if I have to do `state?.route?.query?.current` inside a selector to expose it through my API, that's ok.

Reply:
Both of you have valid points.
For what I can see, the discussion here is about the real meaning of `null`: for @chriskmnds it's “knowingly missing/empty value”, while for @jsnajdr the `null` value “would make sense only when the state value is truly unknown”.

I (as @chriskmnds does) think of `null` as empty value, and `undefined` as truly unknown. If we wanted to respect that, the initial value for `initial`, `previous` and `current` should really be `undefined`.
Given the problem between Redux and `undefined`, and the fact that we're always on some URL, we can say that we can use directly what we intend as `empty value`.

Using `{}` surely has some advantage, for example to use raw selectors throughout the app (like `state.route.query`) without worrying about null values. 

But, for me, it may be worth considering the actual benefit in separation between _data structure_ and _access layer_ using `null` and moving checks and other things up into selectors. Working to make the data consistent in its use of `null` could bring benefits.